### DISCUSSION TITLE
Updating faiss index in queue-like fashion using approximate search

### DISCUSSION THREAD

**Discussion by Leonard907** [facebookresearch/faiss#3672] [Q&A]
# Summary

I have the following use case for faiss: I want to build a index that has fixed size, and I will update the index like a queue (i.e. first in first out). For example, if I want the index to have a bound size of 100 and I already added 100 vectors to it, then if I add another 20 vectors the index should pop out the first 20 vectors.  
The above use case is achievable using the brute-force `IndexFlatL2` index, but not achievable through the IVF index. I have attached an experimental code below. I'm intended to use the approximate search index to do this job as it runs in less time. Do you have any advice on my situation? Thanks in advance!

# Platform

OS: Any 

Faiss version: faiss-gpu 1.7.2 

Installed from: anaconda  

Faiss compilation options: 

Running on:
- [v] CPU
- [ ] GPU

Interface: 
- [ ] C++
- [v] Python

# Reproduction instructions

```
import faiss
import numpy as np

quantizer = faiss.IndexFlatL2(5)
index = faiss.IndexIVFFlat(quantizer, 5, 128)
index.train(np.random.rand(1280, 5).astype('float32'))

for i in range(1000):
    index.add(np.random.rand(100, 5).astype('float32'))
    if index.ntotal > 1280:
        index.remove_ids(np.arange(index.ntotal - 1280))
    d, ind = index.search(np.random.rand(500, 5).astype('float32'), 10)
    if np.max(ind) > 1279:
        import pdb; pdb.set_trace()
    if i % 10 == 9:
        print('Completed: {}/{}'.format(i+1, 1000))
```

**Comment by mdouze** [comment]
That should work. For efficiency, please use Hashtable DirectMap and IDSelectorArray. 

https://github.com/facebookresearch/faiss/wiki/Special-operations-on-indexes#removing-elements-from-an-index

**Comment by Leonard907** [comment]
Is there an example usage of `IDSelectorArray`? I searched in the repo and found some occurrences in test files but I can't quite get its usage. Let's just say I want to remove the first 10 items of an index of size 100, how can I achieve that through IDSelectorArray?

**Comment by mdouze** [comment]
I started adding demo scripts here: 

https://github.com/facebookresearch/faiss/wiki/Case-studies#implementing-an-evolving-ivf-dataset

**Comment by Leonard907** [comment]
@mdouze  Thank you for the demonstration, I can understand the usage now. I experimented with the code a bit and found that the reconstructed index vectors do not match the added vectors exactly. Specifically, if I call `index.reconstruct(0, 1)` and compare it to `xb[0]`, they are not equal to each other. Is this expected? Is there a reason why this is the case?
